nginx/nginx.go

```diff
@@ -0,0 +1,233 @@
+package nginx
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/coreos/ktestutil/utils"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	extensionsv1beta1 "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+)
+
+// Nginx contains information for nginx deployment/service pair.
+type Nginx struct {
+	Namespace string
+	// Name of the deployment
+	Name string
+	// Name of the service
+	SVCName string
+	// List of pods that belong to the deployment
+	Pods []*v1.Pod
+
+	kc       kubernetes.Interface
+	labelSel *metav1.LabelSelector
+}
+
+// NewNginx create this nginx deployment/service pair.
+// It waits until all the pods are running.
+func NewNginx(kc kubernetes.Interface, namespace string) (*Nginx, error) {
+	//create random suffix
+	name := fmt.Sprintf("nginx-%s", utilrand.String(5))
+
+	n := &Nginx{
+		Namespace: namespace,
+		Name:      name,
+		SVCName:   name,
+		labelSel: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"app": name,
+			},
+		},
+		kc: kc,
+	}
+
+	//create nginx deployment
+	nd := n.getDeploymentSpecs()
+	if _, err := kc.ExtensionsV1beta1().Deployments(n.Namespace).Create(nd); err != nil {
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+	if err := utils.Retry(10, 5*time.Second, func() error {
+		d, err := kc.ExtensionsV1beta1().Deployments(n.Namespace).Get(n.Name, metav1.GetOptions{})
+		if err != nil {
```

Comment:
Is this a recoverable error? The retry method you are using will keep retrying this until the # of attempts is reached regardless.

Perhaps consider using apimachinery's `util.PollImmediate` instead?